Even Pods Spread - 1. API changes

Huang-Wei opened the pull request:
**What type of PR is this?**

/kind api-change
/sig scheduling
/hold

**What this PR does / why we need it**:

This is the **1st** PR of the "Even Pods Spread" [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/20190221-even-pods-spreading.md) implementation.

- Manual API changes
- Defaulting, validation and spec removal if feature gate is disabled
- Unit tests
- Auto-generated files

**Special notes for your reviewer:**:

**Which issue(s) this PR fixes**:

Part of #77284.

**Does this PR introduce a user-facing change?**:

```release-note
`[]TopologySpreadConstraint` is introduced into PodSpec to support the "Even Pods Spread" alpha feature.
```

Huang-Wei committed:
EvenPodsSpread: api changes

Huang-Wei commented:
/priority important-soon

/cc @bsalamat @lavalamp @liggitt 
for review.

bsalamat commented on pkg/apis/core/types.go:
```diff
@@ -4755,3 +4761,42 @@ const (
 	DefaultHardPodAffinitySymmetricWeight int32 = 1
 )
+
+type UnsatisfiableConstraintResponse string
```
An integer would be less expensive to compare. Is there any recommendation on the type?

Huang-Wei replied:
I'm ok with `UnsatisfiedContraintAction `. Would like to hear from API experts on naming and the type.

/cc @lavalamp @liggitt 

(BTW: ReadWriteOnce/ReadOnlyMany/ReadWriteMany are of type string)

Huang-Wei replied:
Let's summarize all suggestions and make a change later.

Huang-Wei replied:
It's been renamed to "Action".

bsalamat commented on pkg/apis/core/types.go:
```diff
@@ -2677,6 +2677,12 @@ type PodSpec struct {
 	EnableServiceLinks *bool
+	// TopologySpreadConstraints describes how a group of pods ought to spread across topology
+	// domains. Scheduler respects and enforces the constraints.
+	// This field is alpha-level and is only honored by servers that enables the EvenPodSpread
```
s/servers that enables/clusters that enable/

bsalamat commented on pkg/apis/core/types.go:
```diff
@@ -4755,3 +4761,42 @@ const (
+	// exceeds "MaxSkew".
+	// It's a required field.
+	WhenUnsatisfiable UnsatisfiableConstraintResponse
+	// Label selector for pods. This's enforced by scheduler to check which pods
```
// LabelSelector is used to find matching pods. Pods that match this label selector are counted to determine the number of pods in their corresponding topology domain.

bsalamat commented on pkg/features/kube_features.go:
```diff
@@ -434,6 +434,12 @@ const (
+	// alpha: v1.15
+	//
+	// Schedule pods evenly across available topology domains.
+	EvenPodsSpread utilfeature.Feature = "EvenPodsSpread"
```
Is `EvenPodSpreading` a better name for the feature?

Huang-Wei replied:
I recalled @smarterclayton [mentioned](https://github.com/kubernetes/enhancements/pull/851#discussion_r274250936) that `ing` is not a good name. But I'm open to any reasonable name of the feature gate.

Huang-Wei replied:
Any suggestion on the naming?

cc/ @bsalamat @lavalamp @liggitt @smarterclayton

bsalamat commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5279,3 +5280,34 @@ func ValidateProcMountType(fldPath *field.Path, procMountType core.ProcMountType
+	supportedRespModes := sets.NewString(string(core.DoNotSchedule), string(core.ScheduleAnyway))
+	// tuple {topologyKey, whenUnsatisfiable} denotes one kind of constraint
+	allConstraintKinds := sets.String{}
+	if len(constraints) > 0 {
```
No need for the if statement.

bsalamat commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5279,3 +5280,34 @@ func ValidateProcMountType(fldPath *field.Path, procMountType core.ProcMountType
+func validateTopologySpreadConstraints(constraints []core.TopologySpreadConstraint, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	supportedRespModes := sets.NewString(string(core.DoNotSchedule), string(core.ScheduleAnyway))
```
We should probably build this constant set once and use it later, instead of rebuilding it every time.

lavalamp replied:
+1

Huang-Wei replied:
done.

bsalamat commented on pkg/api/pod/util_test.go:
```diff
@@ -1552,3 +1552,93 @@ func TestDropSubPathExpr(t *testing.T) {
+
+					switch {
+					case enabled || oldPodHasConstraints:
+						// new pod should not be changed if the feature is enabled, or if the old pod // had topology spread constraints
```
nit: fix the comment. it has // in it.

bsalamat commented on pkg/api/pod/util_test.go:
```diff
@@ -1552,3 +1552,93 @@ func TestDropSubPathExpr(t *testing.T) {
+		hasConstraints bool
+		pod            func() *api.Pod
+	}{
+		{
```
Why don't we put a pod, whether the feature is enabled/disabled, and the expected pod in each test case and check the pod coming out of `dropDisabledFields` against the expected pod? I think that would simplify the test.

Huang-Wei replied:
This style is consistent with other tests in this file. I'm not 100% sure about the original intent.

This test is a 2(enabled or not) * 2(old spec with or without constraints) *2(new spec with or without constraints) matrix, I _guess_ it's b/c if we go with the {want, expected} style, we have to give 8 {want, expected} tuples, and the number would increase exponentially if we have other dimensions in the matrix.

Huang-Wei replied:
Hmm.. I'm following the style in this file. So maybe we can come up with a follow-up PR discussing and proposing a new style.

lavalamp replied:
I agree w/ @bsalamat, I don't see the purpose of the "hasConstraints" variable. We shouldn't copy a style if we don't understand its purpose :)

Huang-Wei replied:
I've removed the variable "hasConstraints" as it can be calculated. But remains the "matrix" testing style.

bsalamat reviewed (commented):
Thanks, @Huang-Wei! Most of my comments are minor.

Huang-Wei commented:
Thanks @bsalamat ! Most of the comments have been addressed. PTAL.

lavalamp commented:
/assign

I'll review the API (might not get to it today though)

krmayankk commented on pkg/apis/core/types.go:
```diff
@@ -4767,3 +4773,47 @@ const (
+	// Pods that match this label selector are counted to determine the number of pods
+	// in their corresponding topology domain.
+	// +optional
+	LabelSelector *metav1.LabelSelector
```
i am trying to remember what was the actual use case for the label selector ?

krmayankk replied:
Another way of asking this is why would this label selector be different then the deployments own label selector ? I see two scenarios:-
1: i want to spread all pods of a workload controller (deployment, RS, statefulset). In this case we dont need a selector
2: i want to spread all pods matching a label. This is more common in CRD controllers which deploy a pod and ditch the regular core controllers.
I feel like we need a model where specifying this labelSelector for the first case should not be mandatory and should be derivable from the uber object . @Huang-Wei @bsalamat  wdyt ?

Huang-Wei replied:
A reasonable model is to not impose limitations in the beginning unless we have a strong reason (e.g. perf) to do that.

krmayankk replied:
@Huang-Wei  we discussed this in one of our sig-scheduling meetings to deduce the labelselector from its owning Deployment if one is omitted from the topologySpreadConstraint . This would make it painless for majority of the users who just want to spread the pods of their Deployment /Statefulset and wont have to specify the LabelSelector. I can create a issue on this if you want or if you want to address as part of the other implementation PR's you have.

krmayankk commented on pkg/apis/core/types.go:
```diff
@@ -4767,3 +4773,47 @@ const (
+	// schedule it onto zone1(zone2) will make the ActualSkew(2) violates MaxSkew(1)
+	// - if MaxSkew is 2, incoming pod can be scheduled to any zone.
+	// It's a required value. Default value is 1 and 0 is not allowed.
+	MaxSkew int32
```
should we make it unsigned ?

Huang-Wei replied:
I also thought of this `uint32`. But it seems we're not enforcing a rule - `DeploymentSpec.Replicas` is using `int32`.

krmayankk replied:
any suggestions from @kubernetes/sig-api-machinery-api-reviews  ?

krmayankk commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5279,3 +5280,32 @@ func ValidateProcMountType(fldPath *field.Path, procMountType core.ProcMountType
+	allConstraintKinds := sets.String{}
+	for i, constraint := range constraints {
+		subFldPath := fldPath.Index(i)
+		if val := constraint.MaxSkew; val <= 0 {
```
why do we need to check this , lets make this unsigned

Huang-Wei replied:
with `uint32`, it also needs to be validated to be non-zero.

lavalamp replied:
fyi @krmayankk: Please do not use uint types for restricting the domain to positive integers--it encourages math errors. The code is correct.

krmayankk replied:
@lavalamp  could you elaborate ? underflow/overflow can occur in both . Keeping it unsigned means i dont need to validate if its negative 

@Huang-Wei why should it be validated to be non zero ? It can be zero, which means perfect spreading with no skew

lavalamp replied:
at zero, it will be impossible to legally add another pod. So zero can't be allowed.

underflow can occur always, but underflow at 0 is vastly more likely to be hit in normal code. e.g. `if x - y > z` doesn't do what people expect any more.

Huang-Wei replied:
We can't use zero to describe a "perfect" balanced case - think about how you can spread 2 pods into 2 zones if it's set to zero.

It's similar to the definition of "perfect balanced tree" - where we usually define it as "height of the left and right subtrees from every node differ by 1(maxSkew) or less"

krmayankk replied:
thanks @lavalamp @Huang-Wei for the explanation and it makes sense.  Currently the scheduler decides on a per pod basis .For a hypothetical use case, where an application cannot tolerate unbalance by even one pod, what would be ideal is to allow the scheduler to schedule in batches. So in your example of two pods and two zones, the next scheduling can happen for 2 or 4 more pods otherwise it fails. May be we support this when we have batch scheduling (i think there was some work going on). 

We can even argue, whether we should even design such applications which cannot tolerate an imbalance of 1 pod .

krmayankk commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -13145,3 +13145,101 @@ func TestValidateOrSetClientIPAffinityConfig(t *testing.T) {
+			errfield: "topologyKey",
+		},
+		{
+			name: "duplicate constraints",
```
nit: do you mean two constraints with the same topologyKey ?

Huang-Wei replied:
I meant constraints with duplicated tuple {topologyKey, WhenUnsatisfiable}.

krmayankk reviewed (commented):
added some comments, excited to see this coming

Huang-Wei commented on pkg/api/pod/util_test.go:
```diff
@@ -1552,3 +1552,94 @@ func TestDropSubPathExpr(t *testing.T) {
+						}
+						// new pod should not have subpaths
+						if !reflect.DeepEqual(newPod, podWithoutTopologySpreadConstraints()) {
+							t.Errorf("new pod had subpaths: %v", diff.ObjectReflectDiff(newPod, podWithoutTopologySpreadConstraints()))
```
Thanks for the catch. It's a copy/paste typo...

Huang-Wei commented:
Friendly ping @lavalamp

lavalamp commented:
This is 2nd in my queue. I am overloaded. I may not get to this today.

Huang-Wei commented:
Considering we're close (~2 weeks) to the code freeze of 1.15. Kindly ping @lavalamp again.

lavalamp commented:
Let's look for another reviewer. I am over booked.

*From: *Wei Huang <notifications@github.com>
*Date: *Tue, May 14, 2019 at 7:46 AM
*To: *kubernetes/kubernetes
*Cc: *Daniel Smith, Mention

Considering we're close (~2 weeks) to the code freeze of 1.15. Kindly ping
> @lavalamp <https://github.com/lavalamp> again.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/77327?email_source=notifications&email_token=AAE6BFRLT72P477Q572WY7TPVLGE5A5CNFSM4HJ2Y4K2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVLWUCY#issuecomment-492268043>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAE6BFVP4U5TUBWN2YP4CF3PVLGE5ANCNFSM4HJ2Y4KQ>
> .
>

krmayankk commented on pkg/apis/core/types.go:
```diff
@@ -2686,6 +2686,12 @@ type PodSpec struct {
 	// If not specified, the default is true.
 	// +optional
 	EnableServiceLinks *bool
+	// TopologySpreadConstraints describes how a group of pods ought to spread across topology
```
nit: replace `ought to` with `should be`

lavalamp replied:
The sentence is grammatical as written, please don't change.

krmayankk commented on pkg/apis/core/types.go:
```diff
@@ -2686,6 +2686,12 @@ type PodSpec struct {
 	// +optional
 	EnableServiceLinks *bool
+	// TopologySpreadConstraints describes how a group of pods ought to spread across topology
+	// domains. Scheduler respects and enforces the constraints.
```
nit: `respects this during scheduling`, enforcing sounds like it does something at runtime to enforce it or some other wording ?

krmayankk commented on pkg/apis/core/types.go:
```diff
@@ -2686,6 +2686,12 @@ type PodSpec struct {
 	EnableServiceLinks *bool
+	// TopologySpreadConstraints describes how a group of pods ought to spread across topology
+	// domains. Scheduler respects and enforces the constraints.
+	// This field is alpha-level and is only honored by clusters that enables the EvenPodSpread
```
nit: replace `EvenPodSpread`  with `EvenPodsSpread`

krmayankk reviewed (commented):
did one more pass at the apis

lavalamp commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2982,6 +2982,58 @@ type PodSpec struct {
 	// +optional
 	EnableServiceLinks *bool `json:"enableServiceLinks,omitempty" protobuf:"varint,30,opt,name=enableServiceLinks"`
+	// TopologySpreadConstraints describes how a group of pods ought to spread across topology
+	// domains. Scheduler respects and enforces the constraints.
```
Change the last sentence to "Scheduler respects the constraints during scheduling."

Does another component work to make the constraints true after pods have been scheduled (i.e. by evicting pods)?

Huang-Wei replied:
At this moment, no. We don't proactively balance the pods by evicting pods.

lavalamp commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2982,6 +2982,58 @@ type PodSpec struct {
+
+// TopologySpreadConstraint specifies how to spread matching pods among the given topology.
+type TopologySpreadConstraint struct {
+	// MaxSkew describes the degree of imbalance of pods distribution.
```
```suggestion
	// MaxSkew describes the degree to which pods may be unevenly distributed.
```

lavalamp commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2982,6 +2982,58 @@ type PodSpec struct {
+// TopologySpreadConstraint specifies how to spread matching pods among the given topology.
+type TopologySpreadConstraint struct {
+	// MaxSkew describes the degree of imbalance of pods distribution.
+	// It's the max difference between the number of matching pods in any two
```
```suggestion
	// It's the maximum permitted difference between the number of matching pods in any two
```

lavalamp commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2982,6 +2982,58 @@ type PodSpec struct {
+type TopologySpreadConstraint struct {
+	// MaxSkew describes the degree of imbalance of pods distribution.
+	// It's the max difference between the number of matching pods in any two
+	// topology domains against a given topology type.
```
```suggestion
	// topology domains of a given topology type.
```

lavalamp commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2982,6 +2982,58 @@ type PodSpec struct {
+	// It's the max difference between the number of matching pods in any two
+	// topology domains against a given topology type.
+	// For example, in a 3-zone cluster, currently pods with the same labelSelector
+	// spread as 1/1/0:
```
```suggestion
	// are "spread" such that zone1 and zone2 each have one pod, but not zone3.
```

lavalamp commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2982,6 +2982,58 @@ type PodSpec struct {
+	// For example, in a 3-zone cluster, currently pods with the same labelSelector
+	// spread as 1/1/0:
+	// - if MaxSkew is 1, incoming pod can only be scheduled to zone3 to become 1/1/1;
+	// schedule it onto zone1(zone2) will make the ActualSkew(2) violates MaxSkew(1)
```
```suggestion
	// scheduling it onto zone1(zone2) would make the ActualSkew(2) violate MaxSkew(1)
```

lavalamp commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2982,6 +2982,58 @@ type PodSpec struct {
+	// It's a required field. Default value is 1 and 0 is not allowed.
+	MaxSkew int32 `json:"maxSkew" protobuf:"varint,1,opt,name=maxSkew"`
+	// TopologyKey is the key of node labels. Nodes that have a label with this key
+	// and the value of the label is identical are considered in the same topology.
```
```suggestion
	// and identical values are considered to be in the same topology.
```

Huang-Wei commented:
@lavalamp Thanks for the review! Comments have been addressed. PTAL.

lavalamp commented:
I only made it partway through, sorry, I'll have more comments when I have
another time slice :/

On Tue, May 21, 2019 at 11:10 PM Wei Huang <notifications@github.com> wrote:

> @lavalamp <https://github.com/lavalamp> Thanks for the review! Comments
> have been addressed. PTAL.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/77327?email_source=notifications&email_token=AAE6BFXBSHCYYUIH7DC4Z53PWRQMJA5CNFSM4HJ2Y4K2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODV5GAPI#issuecomment-494559293>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAE6BFXXWB2GBFRWP7L4ARLPWRQMJANCNFSM4HJ2Y4KQ>
> .
>

lavalamp commented on pkg/api/pod/util.go:
```diff
@@ -379,6 +379,10 @@ func dropDisabledFields(
 
+	if !utilfeature.DefaultFeatureGate.Enabled(features.EvenPodsSpread) && !topologySpreadConstraintsInUse(oldPodSpec) {
+		// Set TopologySpreadConstraints to nil only if feature is disabled and it is not used
+		podSpec.TopologySpreadConstraints = nil
```
This doesn't make a lot of sense to me... what is the purpose of setting it to nil if it is not used? Wouldn't that mean it is already nil?

This field is not mutable, which means on a PUT / PATCH request, it cannot be changed at all. This should therefore only apply to POST requests.

Will this get called for deployments, Replica sets, etc, templates?

Huang-Wei replied:
If the feature is disabled, but user still posts a payload with spreadConstraints set, it's set to nil. So basically it's for accidental prevention. I checked the other logic in `dropDisabledFields` function, they're almost using the same pattern. 

> Will this get called for deployments, Replica sets, etc, templates?

Yes, for both Pod and PodTemplate.

krmayankk replied:
i think there was a downgrade/upgrade angle to it as well, i remember someone explaining this, but i have forgotten. I agree whenever i see that piece of code , it doesnt make sense but its a pattern being followed everywhere. I think @liggitt had an explanation

lavalamp replied:
I think you misunderstood my question. the line of code has `!topologySpreadConstraintsInUse(oldPodSpec)`; it seems pointless to set it to nil if it is already not in use. Shouldn't this unconditionally set it to nil?

Huang-Wei replied:
My apologies... I basically agree. I will update it unless hearing from @liggitt about the potential reason.

liggitt replied:
this sets it to nil in the *new* object if the old object did not use the field

liggitt replied:
> Shouldn't this unconditionally set it to nil?

if the feature is enabled, you should be able to update a deployment that does not use this field to start using it

Huang-Wei replied:
Thanks @liggitt.

@lavalamp as Jordan pointed out, this is necessary for k8s version upgrade, or feature gate flipping within the same k8s version. So I will leave the logic as is.

lavalamp replied:
Sorry for the noise, I misread and didn't notice that oldPodSpec and podSpec were two different things.

lavalamp commented on pkg/api/pod/util_test.go:
```diff
@@ -1754,3 +1754,91 @@ func TestDropSubPathExpr(t *testing.T) {
+		},
+	}
+
+	for _, enabled := range []bool{true, false} {
```
This set of nested for loops is MUCH harder to understand than just putting all combos in the table with expected results. Tests with a lot of logic are much harder to verify.

Even if the rest of the file is like this, I would prefer this test to change.

Also, does this need to be tested on things that embed pod templates?

krmayankk replied:
> Also, does this need to be tested on things that embed pod templates?

The dropping of fields will apply to both pod and pod templates and ideally should be tested in both places

Huang-Wei replied:
Done.

lavalamp commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5332,3 +5333,32 @@ func ValidateProcMountType(fldPath *field.Path, procMountType core.ProcMountType
+	allConstraintKinds := sets.String{}
+	for i, constraint := range constraints {
+		subFldPath := fldPath.Index(i)
+		if val := constraint.MaxSkew; val <= 0 {
```
I think generally folks make separate functions rather than validate a bunch of things inside a loop like this.

Huang-Wei replied:
done

lavalamp commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5332,3 +5333,32 @@ func ValidateProcMountType(fldPath *field.Path, procMountType core.ProcMountType
+		if !supportedRespModes.Has(string(constraint.WhenUnsatisfiable)) {
+			allErrs = append(allErrs, field.NotSupported(subFldPath.Child("whenUnsatisfiable"), constraint.WhenUnsatisfiable, supportedRespModes.List()))
+		}
+		kind := fmt.Sprintf("{%s, %s}", constraint.TopologyKey, constraint.WhenUnsatisfiable)
```
it's slightly safer to use a struct{string,string} as a key.

lavalamp replied:
(imagine topology keys with commas in them)

Huang-Wei replied:
done.

lavalamp commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2982,6 +2982,58 @@ type PodSpec struct {
+	// This field is alpha-level and is only honored by clusters that enables the EvenPodsSpread
+	// feature.
+	// +optional
+	// +patchMergeKey=topologyKey
```
According to your validation code, there's actually two keys. Please set the list keys as here: https://github.com/kubernetes/api/blob/kubernetes-1.15.0-alpha.0/core/v1/types.go#L2082-L2095

krmayankk replied:
ah we are missing this 
```
listType=map
	// +listMapKey=containerPort
	// +listMapKey=protocol
``` , wondering how is this used ?

Huang-Wei replied:
Done. Nice to know this "black magic".

lavalamp commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -13279,3 +13279,101 @@ func TestValidateWindowsSecurityContextOptions(t *testing.T) {
+			errfield: "maxSkew",
+		},
+		{
+			name: "missing TopologyKey MaxSkew",
```
omit the trailing "MaxSkew"?

Huang-Wei replied:
done.

lavalamp commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2982,6 +2982,58 @@ type PodSpec struct {
+	// - DoNotSchedule (default) tells the scheduler not to schedule it
+	// - ScheduleAnyway tells the scheduler to still schedule it
+	// Note: it's considered as "Unsatisfiable" only when actual skew on all nodes
+	// exceeds "MaxSkew".
```
...and this condition can't be fixed or improved by adding an additional pod?

e.g. suppose max skew is 1, and the number of pods across 3 topologies is 3/3/1. It seems we should be able to schedule because that would fix the imbalance, right? What if the split is 3/1/1? Then adding 2 pods could fix it. If there are two pending pods, does the system not schedule either of them?

Huang-Wei replied:
In both the 2 cases you gave, we don't see pending pods (unless due to failure on other predicates like node resource shortage). For the split 3/3/1, the incoming pod can be put to topology 3; for the split 3/1/1, the incoming pod can be put to topology 2/3.

One example we will see pending pod is:

```
+---------------+-------+
|     zone1     | zone2 |
+---------------+-------+
| nodeA | nodeB | nodeC |
+-------+-------+-------+
|  P P  |   P   |  P P  |
+-------+-------+-------+
```

Suppose we have two topologySpreadConstraints on both zone and node, with maxSkew 1. In the above example, incoming pod can't be placed:

- if it's put to zone1, the split on **zone** would be imbalanced: 4 vs. 2
- if it's put to zone2(nodeC), the split on **node** would be imbalanced: 2 vs. **1** vs. **3**

lavalamp replied:
Please update the comment.

Huang-Wei replied:
> it's not clear that this permits assignments which improve the skew when it is out of the allowed range

Not exactly. It doesn't improve the skew - it's the "minMatch" that got updated.
For the 3/3/1 case, "minMatch" was 1:
- if we put the incoming pod to topology 3, the global "minMatch" becomes 2, this is allowed.
- if we put the incoming pod to topology 1/2, the global "minMatch" remains 1, so for sure it's not allowed.

lavalamp replied:
My actual question is about the case e.g. 4/4/1. If you have two pods to schedule, I think the behavior people will want is that they both schedule and you end up with 4/4/3. I think the behavior you're describing is that neither pod will schedule. Please make the behavior of this condition clear in the comment.

Also let me urge you to make it permit scheduling pods that don't make things worse: imagine 4/1/1 with 6 pending pods to schedule. IMO the system is unacceptably brittle if the scheduler can't fix things by scheduling the pending pods.

I'm going OOO so I won't be able to continue this conversation, very sorry.

Huang-Wei replied:
> Please make the behavior of this condition clear in the comment.

Will do.

> Also let me urge you to make it permit scheduling pods that don't make things worse: imagine 4/1/1 with 6 pending pods to schedule. IMO the system is unacceptably brittle if the scheduler can't fix things by scheduling the pending pods.

Exactly. The internal implementation is exactly as what you said - the only possible final result is from 4/1/1 to 4/4/4.

> I'm going OOO so I won't be able to continue this conversation, very sorry.

It's totally fine as we will target this to 1.16. And during your absence, I will request reviewers to review afterward PRs which are only related with scheduling.

Appreciate your efforts reviewing this!

Huang-Wei replied:
Comments updated.

lavalamp commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2982,6 +2982,58 @@ type PodSpec struct {
+	// Note: it's considered as "Unsatisfiable" only when actual skew on all nodes
+	// exceeds "MaxSkew".
+	// It's a required field.
+	WhenUnsatisfiable UnsatisfiableConstraintAction `json:"whenUnsatisfiable" protobuf:"bytes,3,opt,name=whenUnsatisfiable,casttype=UnsatisfiableConstraintAction"`
```
Is there a policy for what to do if the pods become imbalanced via attrition?

krmayankk replied:
@lavalamp this Even Pod Spreading feature only applies to scheduling time decisions. Runtime even spreading due to any number of issues like drain, auto scaling ,etc is not covered. I am making a change in descheduler to incorporate the same policy for runtime. That means, when maxSkew is not honored for a topokey, the descheduler will evict some pods and assume the scheduling will do the right thing.

Having said, i would be interested in knowing if container orchestrators like . Borg, Tupperware handle runtime spreading ? Would that be something we could incoporate into k/k ?

lavalamp replied:
If a descheduler is expected to take action on this as well, the comments should document that behavior. (Although that seems a little weird to me, I'd expect this to only be relevant if we are scaling down a deployment. If e.g. 2 pods in one zone fail, I'd expect users to be quite angry if the descheduler then killed 1 pod in each of their other zones!!)

Huang-Wei replied:
> Is there a policy for what to do if the pods become imbalanced via attrition?

Default scheduler does nothing on it - like what we are doing all the time.

However, @lavalamp you're right, the only case I do want to achieve is the scale-down case - arbitrary taking out a replica is not an ideal case - I'm going to tackle it down in Beta either in scheduler or workload controller side.

krmayankk replied:
@lavalamp  failed pods will not be accounted for in descheduler. Descheduler will only look at evictable pods , which means pods that are running and ready and are not daemonset pods. 

We probably need to document this better, but even pod spreading is only applied when a new pod is scheduled, it should be as far apart in the domain as possible, to other already running pods

Descheduler policy, will look at already running pods, divide them into buckets and see if the spread doesnt follow maxSkew, start evicting some pods from more full buckets and assume scheduling will fix them

@bgrant0607 may be has insight on the Borg question, Does it handle runtime spreading across domains ?

lavalamp replied:
Please answer the question in the comments :)

krmayankk replied:
@lavalamp probably i misunderstood your question. I think attrition can only happen by scaling down. The scaling down case can be solved in two ways

1: fix the scaling down in workload controller to account for topo constraints. I think this is less elegant and spreads the constraint knowledge in both scheduler and workload controller
2: let the descheduler handle the eviction and the scheduler will automagically fix this. although this does require dependency on an external component

>  If e.g. 2 pods in one zone fail, I'd expect users to be quite angry if the descheduler then killed 1 pod in each of their other zones!!)

What do you mean by 2 pods in one zone fail ? Do you mean 2 pods are pending ?

lavalamp commented:
OK, I've made it all the way through. Thanks for your patience.

bsalamat commented:
Thanks, @Huang-Wei for rebasing and addressing comments.
@lavalamp @liggitt Is there anything else that needs to be addressed here? Our other PRs to implement this feature depend on this one. It would be great if we could get this approved sooner.

Huang-Wei commented:
https://github.com/kubernetes/kubernetes/pull/77327#issuecomment-501614011 seems to be a prow bug.

tedyu commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5397,3 +5398,78 @@ func ValidateProcMountType(fldPath *field.Path, procMountType core.ProcMountType
+	allErrs := field.ErrorList{}
+
+	var existingConstraintPairs []spreadConstraintPair
+	for i, constraint := range constraints {
```
Should we check length of constraints matches that of fldPath.Index ?

Huang-Wei replied:
We don't need to.

tedyu commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5397,3 +5398,78 @@ func ValidateProcMountType(fldPath *field.Path, procMountType core.ProcMountType
+
+// ValidateSpreadConstraintPair tests that if `pair` exists in `existingConstraintPairs`.
+func ValidateSpreadConstraintPair(fldPath *field.Path, pair spreadConstraintPair, existingConstraintPairs []spreadConstraintPair) *field.Error {
+	for _, existingPair := range existingConstraintPairs {
```
It seems if existing constraints are represented with a map, the check would become O(1)

Huang-Wei replied:
In theory yes, but in practice we usually have a couple of constraints, so the complexity would outweigh the benefits.

lavalamp commented on pkg/apis/core/types.go:
```diff
@@ -4805,3 +4812,64 @@ const (
+	// MaxSkew(1). In other words, the cluster can still be imbalanced, but scheduler
+	// won't make it *more* imbalanced.
+	// It's a required field.
+	WhenUnsatisfiable UnsatisfiableConstraintAction
```
I'd still love to see some discussion in the comments of how this affects the pod *after* it's been scheduled. E.g. does it affect podDisruptionBudget? (I assume not at the moment but will it ever?) Descheduler? Scale-down behavior?

bsalamat replied:
@lavalamp  We could add more information on how it could affect later scheduling related decisions. Particularly, autoscaler simulates whether a pod being removed from a node can be scheduled on another node during scale-down. This feature may prevent a pod from being scheduled on a different node/zone/etc if the scheduling of the pod breaks the balanced spreading. Therefore, it could prevent scale-down in certain scenarios. This is actually a positive benefit. I have already heard from some users that the autoscaler sometimes aggressively removes nodes in clusters, causing a few pods of a service to be scheduled on a single node. This could reduce availability/reliability of a service if the node with a few pods of the service goes down.
This exact topic came up in our SIG Scheduling F2F meeting at KubeCon where people asked for a "required" spreading feature in order to prevent autoscaler from removing nodes.

All being said, I believe the right place for providing such information about this feature is in its documents. We cannot put all the information in doscstring of the API fields.

Huang-Wei replied:
Once a pod gets scheduled, default scheduler won't touch it any more. This is aligned with our design rationale, as re-scheduling may cause service down time, so we usually suggest users use external tool like descheduler to balance exsiting workloads proactively.

podDisruptionBudget is respected upon preemption, and that's covered in preeemption logic specifically.

Scale-down may be the real missing piece in this feature, at this moment we're relying on controller to take out excessive replicas, but it might not be optimal as controller doesn't have enough info to make that decision.

lavalamp replied:
I think my main point here is that people reading these api comments will not only care about the scheduler, they will care about the entire lifecycle of their pods / services. I think it's reasonable to at least give a breadcrumb here to other documentation to learn about how the system behaves after these pods have been scheduled.

Huang-Wei replied:
It makes sense. I will include a link here once the doc is baked.

krmayankk replied:
When we say scale down are we talking about using the scale subresource to reduce the replica count of a Deployment or otherwe resources ?For that we would need logic in the deployment controller to say bring down a pod , which doesnt affect even spreading

Huang-Wei replied:
@krmayankk you're right.

lavalamp commented:
I think my comments have been addressed, so this is an API LGTM.

Huang-Wei commented:
/retest

krmayankk commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5405,3 +5406,78 @@ func ValidateProcMountType(fldPath *field.Path, procMountType core.ProcMountType
+func ValidateSpreadConstraintPair(fldPath *field.Path, pair spreadConstraintPair, existingConstraintPairs []spreadConstraintPair) *field.Error {
+	for _, existingPair := range existingConstraintPairs {
+		if pair.topologyKey == existingPair.topologyKey &&
+			pair.scheduleAction == existingPair.scheduleAction {
```
i am not following , it seems you are suggesting we can only have a unique pair so topoKey1,scheduleAction1 and topoKey1,scheduleAction2 are valid pairs ? What would that mean in implementation ? I think if we have topology key repeated that is an error , and  we dont need to check for schedule action @Huang-Wei

Huang-Wei replied:
Two constraints with same {topologyKey, scheduleAction} should be mergeable:

- greater maxSkew should be respected
- labelSelectors are also mergeable

So for the sake of calculation efficiency and conciseness of API, I do believe we should only allow different constraints with different {topologyKey, scheduleAction} tuples.

krmayankk replied:
IMO, it makes it simpler to reason about if we don't allow duplicate topologyKey. We should explicitly flag this as an error rather than silently behind the scenes merge the constraints and leave the user with a behavior he/she didnt expect. If the user wanted something that needs to be merged, he can specify a merged constraint. I see this as more of better usability

Huang-Wei replied:
The merge I mean is patch merging in API side, not that 2 constraints with same topokey+scheduleAction in an API request will be merged - which will fail.

lavalamp replied:
Duplicate key makes lots of sense if you prefer skews below X, and can't tolerate skews above Y, X < Y.

krmayankk replied:
interesting Idea, thanks for bringing that up @lavalamp . Its too early to say we want this, but lets say we want to implement this, then we have two options:-
```
**option1:** 
{ 
   topologyKey: topo1
   maxSkew: 3
} 
{ 
   topologyKey: topo1
   maxSkew: 5
} 


**option2:** 
{ 
   topologyKey: topo1
   skewRange: {
    min: 3
    max: 5
  }  
```

I think option 2 is much more clear and user friendly. WDYT @lavalamp @bsalamat @Huang-Wei ?


@Huang-Wei the UT you have says if you find duplicate tuple(topoKey and scheduleAction), you will fail. That means your validation will pass for tuples with same topoKey and different scheduleActions. 

I am suggesting if we decide to not implement the idea by @lavalamp , then we should not allow tuples with same topoKey and different scheduleActions. It will be difficult to document what that behavior means . We should simplify this and say two constraints with same topoKey dont make any sense and we should error out .

Huang-Wei replied:
@krmayankk I'm not sure if we're talking about the same thing. Both @lavalamp and I are just talking about the following combination can be possible (where X<Y):

1. topologyKey=zone, scheduleAction=ScheduleAnyway, maxSkew=X - it's a **soft** requirement, i.e. it's a good-to-have
1. topologyKey=zone, scheduleAction=DoNotSchedule, maxSkew=Y - it's a **hard** requirement, i.e. it's a must-have

In above case, the 2nd constraint MUST be obeyed, and the 1st constraint is just _preferred_.

---

The `skewRange` thing you raised is confusing b/c when it comes to term "range", it reminds me of the formula "min < actual skew < max", which is absolutely not the same thing "even pods spread" is talking about.

krmayankk replied:
I see your point. Is your subsequent PR's implementing this logic already ?

The range was just an example,We could call it something else, but the point is , the user should not have to repeat the topologySpreadConstraint, if they just need to specify a soft and a hard requirement with different skew.  So it could look like this:
```
   topologyKey=zone
   softSkew: X
   hardSkew: Y
```
One of softSkew or hardSkew must be specified. I think this is easier and more user friendly representation then 
```
{
    topologyKey=zone
    scheduleAction=ScheduleAnyway
    maxSkew=Y
},
{
    topologyKey=zone
    scheduleAction= DoNotSchedule
    maxSkew=Y
}
```
Its an alpha so we can get feedback from the community, but I have a hard time thinking why someone would choose the second option over the first one .

lavalamp replied:
Please leave this the way it is. Putting the action in fields (whether min/max or soft/hard) like this makes it much harder to add additional constraint types in the future.

Huang-Wei replied:
@krmayankk I see your point. But I agree more with @lavalamp - to make the options standalone (instead of mixing them in the beginning) make it clear and extensible.

krmayankk commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -13499,7 +13498,104 @@ func TestAlphaVolumePVCDataSource(t *testing.T) {
+			name: "duplicate constraints",
+			constraints: []core.TopologySpreadConstraint{
+				{MaxSkew: 1, TopologyKey: "k8s.io/zone", WhenUnsatisfiable: core.DoNotSchedule},
+				{MaxSkew: 2, TopologyKey: "k8s.io/zone", WhenUnsatisfiable: core.DoNotSchedule},
```
Why would we have two topologyKey exactly same in the same TopologySpreadConstraint ?

Huang-Wei replied:
This is an intentional failure case.

krmayankk commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -13499,7 +13498,104 @@ func TestAlphaVolumePVCDataSource(t *testing.T) {
+		{
+			name: "invalid MaxSkew",
+			constraints: []core.TopologySpreadConstraint{
+				{MaxSkew: 0, TopologyKey: "k8s.io/zone", WhenUnsatisfiable: core.DoNotSchedule},
```
Also maxSkew as 0 is perfectly valid use case for perfect spreading ? Why do you disallow it ?

lavalamp replied:
As stated previously-- at zero, it becomes illegal to ever add an additional pod; imagine (3,3,3): an additional pod anywhere will increase the skew to 1 and therefore be illegal.

Huang-Wei replied:
@krmayankk see our discussion at https://github.com/kubernetes/kubernetes/pull/77327#discussion_r289450609

krmayankk commented:
thanks fire tirelessly working on this. LGTM except for this use case where we are allowing duplicate topologyKey.

Huang-Wei commented:
> LGTM except for this use case where we are allowing duplicate topologyKey.

"Duplicate topologyKey" you mean duplicate topologyKeys only or duplicate {topologyKey, scheduleAction} tuples?

krmayankk commented:
> > LGTM except for this use case where we are allowing duplicate topologyKey.
> 
> "Duplicate topologyKey" you mean duplicate topologyKeys only or duplicate {topologyKey, scheduleAction} tuples?

Duplicate topologyKey only, no matter what other values are should not be allowed

ahg-g commented:
One more thing, and I am not sure if this has been discussed before: looking at the [implementation](https://github.com/kubernetes/kubernetes/pull/77828/files/e6546ee07fdf7cb1833d408d09073035a4835112..4e88e35dfe502a1c2f9103e48699447f7e77dca0#diff-10f83a56e52ed3dac0a5c12727a7156aR1766) in subsequent PRs, I think we don't treat failure domains as a hierarchy, if that is by design, then we should probably document it.

For example, in the following setup
```
{
    topologyKey=zone
    scheduleAction=DoNotSchedule
    maxSkew=1
},
{
    topologyKey=node
    scheduleAction= DoNotSchedule
    maxSkew=1
}
```

It seems that maxSkew for the second constraint will be evaluated across nodes in all zones, not nodes within a zone. The implication of this can be explained in the following example, consider a cluster in the following state:

```
+-------------------+-------------------+
|       zone1       |       zone2       | 
+-------------------+-------------------+
| node1a |  node1b  | node2a |  node2b  |
+--------+----------+--------+----------+
| 6 pods |  1 pods  | 3 pods |  2 pods  |
+--------+----------+--------+----------+
```

Please correct me if I am wrong, but the pod will be rejected because in the [currently proposed implementation](https://github.com/kubernetes/kubernetes/pull/77828/files/e6546ee07fdf7cb1833d408d09073035a4835112..4e88e35dfe502a1c2f9103e48699447f7e77dca0#diff-10f83a56e52ed3dac0a5c12727a7156aR1766): the first constraint will filter ```node1a``` and ```node1b```, while the second constraint will filter ```node2a``` and ```node2b```.

lavalamp commented:
@ahg-g That's a great point, I think if it's implemented as "it's allowed unless it makes the skew worse" it should function? I think that's what the documentation says, at least I think I asked for it to say that :)

Is it possible to come up with an example like yours, but where a placement on any node makes some skew worse? I think it is possible if you e.g. have nodes which are in multiple topology zones, then you can make some cycle. I am pretty sure we'll consider that particular arrangement to be user error, but a) how does the user learn that they've made a mistake, and b) are there other more reasonable (non-cyclical) topology arrangements that also have this property?

Huang-Wei commented:
@ahg-g There is no concept describing "hierarchy" among topology domains b/c the topology domain is nothing but labels applied to nodes, so they can be any valid {key:value} pair. We don't have extra metadata to know which label pair is subjected to which. So right now the logic examines each topologySpreadconstrain, and make the result ANDed. 

The case you pointed out is by design - which has been discussed in our weekly meeting. I've included some UTs to cover this case, and will highlight it in the document.

@lavalamp The comment "it's allowed unless it makes the skew worse" has been added. Regarding the example, a minimized one can be a deployment (replicas=6) is trying to be deployed onto:

```
+-------------------+---------+
|       zone1       |  zone2  |
+-------------------+---------+
| node-1a | node-1b | node-2a |
+---------+---------+---------+
|         |         |         |
+---------+---------+---------+
```

with constraints on both zone and node, and maxSkew for each.

In this case, the **6th** pod won't be placed as it at least violates one constraint:

```
+-------------------+---------+
|       zone1       |  zone2  |
+-------------------+---------+
| node-1a | node-1b | node-2a |
+---------+---------+---------+
|   P P   |    P    |   P P   |
+---------+---------+---------+
```

---

In general, this is by design and has been discussed with @bsalamat. We will document it on the website.

ahg-g commented:
It is true that we don't have a notion of hierarchy (may be we should have one since in reality a hierarchy exist, but that is a discussion for another issue), documenting those cases on the website sounds good to me.

Just something to think about, and I am no API expert: does it make sense to consider the order of the constraints as a way to avoid those cases: basically we compute the min skew over the nodes left after a previous constraint was evaluated. It may not be simple to implement though...

/lgtm

Huang-Wei commented:
> Just something to think about, and I am no API expert: does it make sense to consider the order of the constraints as a way to avoid those cases: basically we compute the min skew over the nodes left after a previous constraint was evaluated. It may not be simple to implement though...

This is an interesting point. Ordering can save efforts as a failure on 1st constraint will short circuit the resting ones. To go a step further, imposing a semantics like "compute the min skew over the nodes left after a previous constraint was evaluated" sounds amazing, but isn't practical b/c it only works if the latter topology is **node**; if the latter one is zone/region/rack, failed nodes in the previous constraint may still contribute to the "number of matching pods" on current zone/region/rack, so we still need to go through them.

Huang-Wei commented:
All in all, how we view a Kubernetes cluster's topology is dependent on labels of nodes. Labels are simple and flexible, but it's not easy/feasible to build a hierarchical tree/graph which makes some of our feature/implementation restrictive.

Huang-Wei commented:
/retest

krmayankk commented:
> One more thing, and I am not sure if this was discussed before: looking at the [implementation](https://github.com/kubernetes/kubernetes/pull/77828/files/e6546ee07fdf7cb1833d408d09073035a4835112..4e88e35dfe502a1c2f9103e48699447f7e77dca0#diff-10f83a56e52ed3dac0a5c12727a7156aR1766) in subsequent PRs, I think we don't treat failure domains as a hierarchy, if that is by design, then we should probably document it.
> 
> For example, in the following setup
> 
> ```
> {
>     topologyKey=zone
>     scheduleAction=DoNotSchedule
>     maxSkew=1
> },
> {
>     topologyKey=node
>     scheduleAction= DoNotSchedule
>     maxSkew=1
> }
> ```
> 
> It seems that maxSkew for the second constraint will be evaluated across nodes in all zones, not nodes within a zone. The implication of this can be explained in the following example, consider a cluster in the following state:
> 
> ```
> +-------------------+-------------------+
> |       zone1       |       zone2       | 
> +-------------------+-------------------+
> | node1a |  node1b  | node2a |  node2b  |
> +--------+----------+--------+----------+
> | 6 pods |  1 pods  | 3 pods |  2 pods  |
> +--------+----------+--------+----------+
> ```
> 
> Please correct me if I am wrong, but the pod will be rejected because in the [currently proposed implementation](https://github.com/kubernetes/kubernetes/pull/77828/files/e6546ee07fdf7cb1833d408d09073035a4835112..4e88e35dfe502a1c2f9103e48699447f7e77dca0#diff-10f83a56e52ed3dac0a5c12727a7156aR1766): the first constraint will filter `node1a` and `node1b`, while the second constraint will filter `node2a` and `node2b`.

This is exactly what i brought up in the meeting earlier, but didnt find a good example. I think in the long term we need to think about this. Since as you said , in reality the domains have a hierarchy

krmayankk commented:
/lgtm

Huang-Wei committed:
EvenPodsSpread: auto-gen files

Huang-Wei committed:
EvenPodsSpread: regenerated API compatibility data

- UPDATE_COMPATIBILITY_FIXTURE_DATA=true go test ./vendor/k8s.io/api -run //HEAD &>/dev/null

Huang-Wei commented:
The recent force-push didn't change the code but just rebasing.

@lavalamp the resting PRs have got reviewed by @ahg-g and @bsalamat, we just have a few nits to figure out. It'd good if you can /approve this one, so that we can merge the resting ones one by one.

Thanks for the tremendous reviewing efforts all guys put on this PR. Really appreciate that!

bsalamat commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5480,3 +5481,78 @@ func ValidateProcMountType(fldPath *field.Path, procMountType core.ProcMountType
+
+type spreadConstraintPair struct {
+	topologyKey    string
+	scheduleAction core.UnsatisfiableConstraintAction
```
+1. It'd make it more consistent with the API, etc.

Huang-Wei replied:
Done.

bsalamat reviewed (commented):
Only a minor comment. Otherwise, LGTM.

/retest

Huang-Wei commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5480,3 +5481,78 @@ func ValidateProcMountType(fldPath *field.Path, procMountType core.ProcMountType
+}
+
+// ValidateScheduleAction tests that the argument is a valid UnsatisfiableConstraintAction.
+func ValidateScheduleAction(fldPath *field.Path, action core.UnsatisfiableConstraintAction) *field.Error {
```
Done.

Huang-Wei commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5480,3 +5481,78 @@ func ValidateProcMountType(fldPath *field.Path, procMountType core.ProcMountType
+}
+
+// ValidateSpreadConstraintPair tests that if `pair` exists in `existingConstraintPairs`.
+func ValidateSpreadConstraintPair(fldPath *field.Path, pair spreadConstraintPair, existingConstraintPairs []spreadConstraintPair) *field.Error {
```
> And you should be able to iterate through the original constraints slice instead of creating a new array of pairs (up to the current index).

Not really. We shouldn't re-iterate the visited constraints.

alculquicondor replied:
Why not? It would be just to check for repetitions. The complexity is the same. Unless you plan to replace []spreadConstraintPair by a map in the future. But we don't expect to have a big amount of constraints.

Huang-Wei replied:
> Complexity is basically the same

Not for the worst/average case.

If we check repeatedly, for the ith constraint, each time it's compared with **_all_** constraints. So in total it's `n^2` times, and notice that we need additional logic to exclude the case it compares with itself.
If we only check **_visited_** constraints, in total it's `1+2+...+n` = `n^2 / 2`. And we don't need to consider the case it compares with itself.

alculquicondor replied:
You just need to pass a subslice: `constraints[:i]`

Huang-Wei commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -13576,7 +13575,104 @@ func TestAlphaVolumePVCDataSource(t *testing.T) {
+		{
+			name: "missing scheduling mode",
+			constraints: []core.TopologySpreadConstraint{
+				{MaxSkew: 1, TopologyKey: "k8s.io/zone"},
```
That's a private variable, so not worth making it public and make a reference here.

alculquicondor replied:
Oh, I missed that. There is still `v1.LabelZoneFailureDomain` or simply a constant at the top of this file. But again, just a nit.

Huang-Wei commented:
@bsalamat @alculquicondor The naming comment addressed. PTAL.

Huang-Wei commented:
/retest

alculquicondor commented:
Hopefully others don't disagree, but it's useful if you add more commits instead of amending. We can squash once the review is finalized.

bsalamat reviewed (commented):
/lgtm
/approve

Thanks a lot, @Huang-Wei for your exemplary dedication and great work!

lavalamp commented:
/approve

Huang-Wei commented:
/hold cancel

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

krmayankk commented:
this is awesome thanks @Huang-Wei  @bsalamat and @lavalamp for getting this merged
